src/functions.jl

```diff
@@ -6,649 +6,479 @@
+        ## acos ##
+        function acos(a::$T)
+            a0 = constant_term(a)
+            a0^2 == one(a0) && throw(ArgumentError(
+                """
+                Recursion formula diverges due to vanishing `sqrt`
+                in the denominator.
+                """))
+
+            order = max_order(a)
+            c = $T( asin(a0), order )
+            r = $T( sqrt(1 - a0^2), order )
+            @inbounds for k in 1:order
+                acos!(c, a, r, k)
+            end
+            @inbounds c[1] = acos(a0)
+            return c
+        end
 
-\begin{equation*}
-c_k = \frac{1}{a_0} (a_k - \frac{1}{k} \sum_{j=0}^{k-1} j a_{k-j} c_j ),
-\end{equation*}
+        ## atan ##
+        function atan(a::$T)
+            order = max_order(a)
+            a0 = constant_term(a)
+            c = $T( atan(a0), order)
+            r = $T(1 + a0^2, order)
+            constant_term(r) == zero(constant_term(a)) &&
+                throw(ArgumentError(
+                    """
+                    Recursion formula has a pole.
+                    """))
+
+            @inbounds for k in 1:order
+                atan!(c, a, r, k)
+            end
+            return c
+        end
 
-with $a$ a `Taylor1` polynomial.
+        ## sinh and cosh ##
+        sinh(a::$T) = sinhcosh(a)[1]
+        cosh(a::$T) = sinhcosh(a)[2]
+        function sinhcosh(a::$T)
+            order = max_order(a)
+            a0 = constant_term(a)
+            s = $T( sinh(a0), order)
+            c = $T( cosh(a0), order)
+            @inbounds for k = 1:order
+                sinhcosh!(s, c, a, k)
+            end
+            return s, c
+        end
 
-Inputs are the `kcoef`-th coefficient, the vector of the expansion coefficients
-`ac` of `a`, and the already calculated expansion coefficients `coeffs` of `c`.
-"""
-function logHomogCoef{T<:Number}(kcoef::Int, ac::Array{T,1}, coeffs::Array{T,1})
-    kcoef == 0 && return log( ac[1] )
-    coefhomog = zero(T)
-    @inbounds for i = 1:kcoef-1
-        coefhomog += (kcoef-i) * ac[i+1] * coeffs[kcoef-i+1]
+        ## tanh ##
+        function tanh(a::$T)
+            order = max_order(a)
+            @inbounds aux = tanh( constant_term(a) )
+            c = $T( aux, order)
+            c2 = $T( aux^2, order)
+            @inbounds for k = 1:order
+                tanh!(c, a, c2, k)
+            end
+            return c
+        end
     end
-    @inbounds coefhomog = (ac[kcoef+1] -coefhomog/kcoef) / ac[1]
-    coefhomog
 end
 
 
-### TRIGONOMETRIC FUNCTIONS ###
-
-## Sin ## 
-doc"""
-    sin(a)
+# Recursive functions (homogeneous coefficients)
+for T in (:Taylor1, :TaylorN)
+    @eval begin
+        function exp!(c::$T, a::$T, k::Int)
+            if k == 0
+                @inbounds c[1] = exp(constant_term(a))
+                return nothing
+            end
 
-Return the Taylor expansion of $\sin(a)$, of order `a.order`, for
-`a::Taylor1` polynomial.
+            c[k+1] = zero_korder(c, k)
```

Comment:
It avoids the nasty `if $T == Taylor1 ... end` block that were there. For `Taylor1` it simply returns `zero(a[1])`, while for `TaylorN` it returns `HomogeneousPolynomial(zero(a[1][1]),k`.